openedx/core/djangoapps/lang_pref/middleware.py

```diff
@@ -21,17 +26,46 @@ class LanguagePreferenceMiddleware(object):
     def process_request(self, request):
         """
         If a user's UserPreference contains a language preference, use the user's preference.
+        Save the current language preference cookie as the user's preferred language.
         """
-        languages = released_languages()
-        system_released_languages = [seq[0] for seq in languages]
+        cookie_lang = request.COOKIES.get(settings.LANGUAGE_COOKIE, None)
+        if cookie_lang:
+            if request.user.is_authenticated():
+                set_user_preference(request.user, LANGUAGE_KEY, cookie_lang)
```

Comment:
The general thing to consider in the cookie vs preference question is: we're going to let the user change their preference via the marketing site, either by selecting a language from a menu, or simply by visiting the spanish landing page. In either of those cases, we don't want to have to actually hit the LMS right away (we're fine w/ updating the stored preference whenever they finally make it to the LMS).